### PR TITLE
rewrite `Interval.parse(CharSequence)`

### PR DESCRIPTION
**Pull request by perceptron8**
This is an attempt to make `Interval` parsing more readable and (hopefully) less error prone. I believe that proposed way of splitting code into methods makes much more sense than before. At least one bug was fixed. At the same time, parsing became a little more flexible.

---

The bug was caused by the following:

```java
long move = end.isBefore(Instant.EPOCH) ? 1000 * 86400 : -1000 * 86400;
Instant start = end.plusSeconds(move).atOffset(ZoneOffset.UTC).minus(amount).toInstant().minusSeconds(move);
```

```java
long move = start.isBefore(Instant.EPOCH) ? 1000 * 86400 : -1000 * 86400;
Instant end = start.plusSeconds(move).atOffset(ZoneOffset.UTC).plus(amount).toInstant().minusSeconds(move);
```

It looks like author's intention was to move instant 1000 days towards `Instant.EPOCH`. First, it is not exactly 1000 days - it's approximately 1000 days. What's important is that it can (but doesn't have to) move instant into month with different number of days. I decided to use leap year cycle length (which is fixed) to avoid this kind of problems.

Without this fix, not a single test covered by `org.threeten.extra.TestInterval.data_parse_outside_bounds(String, String, String)` is going to pass. For instance `-1000000000-01-31T00:00:00Z/P1M` should be parsed to interval ending at `-02-29`, not `-03-02` and for `-1000000000-01-31T00:00:00Z/P2M` interval's end should be `-03-31` instead of `-04-01` .

There are still some cases that this rewrite doesn't fix. All of them involve duration. Please, see `org.threeten.extra.TestInterval.data_parse_crossing_bounds(String, String, String)` (intervals starting within date-time bounds, with duration making them end outside of it) and `org.threeten.extra.TestInterval.data_parse_crossing_bounds_twice(String, String, String)` (intervals starting before `OffsetDateTime.MIN` and ending after `OffsetDateTime.MAX` because of duration). Maybe someone will be interested in making these tests pass - for now, they are disabled.

---

As I mentioned, parsing became a little more flexible. It is now possible to parse `OffsetDateTime/LocalDateTime` and `LocalDateTime/OffsetDateTime` (previously only `OffsetDateTime/LocalDateTime` was supported). It seems to be valid according to spec, at least if Wiki can be trusted.

 > For <start>/<end> expressions, if any elements are missing from the end value, they are assumed to be the same as for the start value including the time zone
(https://en.wikipedia.org/wiki/ISO_8601#Time_intervals)

---

Probably related: #80.

### CODE REVIEW

**Comment by perceptron8** [comment]
LocalDateTime/OffsetDateTime format is illegal again. Sorry for the confusion!

File: src/test/java/org/threeten/extra/TestInterval.java [hunk 1]
```diff
@@ -267,6 +269,39 @@ public void test_parse_CharSequence(String input, Instant start, Instant end) {
...
+        "-1000000000-01-31T00:00:00Z/P1M, -1000000000-01-31T00:00:00Z, -1000000000-02-29T00:00:00Z",
+        "-1000000000-01-31T00:00:00Z/P2M, -1000000000-01-31T00:00:00Z, -1000000000-03-31T00:00:00Z",
+        "P5M/+1000000000-12-31T23:59:59.999999999Z, +1000000000-07-31T23:59:59.999999999Z, +1000000000-12-31T23:59:59.999999999Z",
+        "P10M/+1000000000-12-31T23:59:59.999999999Z, +1000000000-02-29T23:59:59.999999999Z, +1000000000-12-31T23:59:59.999999999Z",
+    })
+    public void data_parse_outside_bounds(String interval, String start, String end) {
+        assertEquals(Interval.of(Instant.parse(start), Instant.parse(end)), Interval.parse(interval));
+    }
+
+    @Disabled("not implemented yet")
```
Comment (jodastephen) [inline src/test/java/org/threeten/extra/TestInterval.java:283]: Does this indicate that the PR is not ready yet?

File: src/main/java/org/threeten/extra/Interval.java [hunk 2]
```diff
@@ -180,80 +186,94 @@ public static Interval parse(CharSequence text) {
...
         for (int i = 0; i < text.length(); i++) {
             if (text.charAt(i) == '/') {
-                return parseSplit(text.subSequence(0, i), text.subSequence(i + 1, text.length()));
+                CharSequence startStr = text.subSequence(0, i);
+                CharSequence endStr = text.subSequence(i + 1, text.length());
+                if (startStr.charAt(0) == 'P' || startStr.charAt(0) == 'p') {
+                    // duration followed by temporal
+                    TemporalAmount duration = parseDuration(startStr);
+                    Temporal temporal = parseTemporal(endStr);
+                    return Interval.of(Instant.from(minus(temporal, duration)), Instant.from(temporal));
```
Comment (jodastephen) [inline src/main/java/org/threeten/extra/Interval.java:195]: I wonder if situations like this should "saturate" the instant at Instant.MIN/MAX, to simplify the lives of callers

Comment (perceptron8) [reply src/main/java/org/threeten/extra/Interval.java:195]: I don't think that's a good idea. First, `Instant` itself doesn't saturate. Both of the following throw `java.time.DateTimeException: Instant exceeds minimum or maximum instant`.

```java
Instant.MIN.minusNanos(1);
```

```java
Instant.MAX.plusNanos(1);
```

Same for `OffsetDateTime.MIN` / `OffsetDateTime.MAX`, different exception.

What's more, `LocalDateRange` (which is somewhat similar to `Interval`) doesn't saturate either. The following ends up with `java.time.DateTimeException: Invalid value for Year (valid values -999999999 - 999999999): 1000000000`.

```java
LocalDateRange.of(LocalDate.MAX.minusDays(2), Period.ofDays(3));
```

In my opinion, any kind of "saturation" would be very surprising to callers.
Let's preserve the status quo.

File: src/test/java/org/threeten/extra/TestInterval.java (hunk 1 above)
Comment (perceptron8) [reply src/test/java/org/threeten/extra/TestInterval.java]: I removed "yet". This PR doesn't fix, but also doesn't break anything related to these cases. These 2 disabled tests are here to make others aware of implementation limitations that are present now and were present before, but are / were not so obvious at first sight.

> There are still some cases that this rewrite doesn't fix. All of them involve duration. Please, see `org.threeten.extra.TestInterval.data_parse_crossing_bounds(String, String, String)` (intervals starting within date-time bounds, with duration making them end outside of it) and `org.threeten.extra.TestInterval.data_parse_crossing_bounds_twice(String, String, String)` (intervals starting before OffsetDateTime.MIN and ending after `OffsetDateTime.MAX` because of duration). Maybe someone will be interested in making these tests pass - for now, they are disabled.

I've tried to implement such cases for a while, but code quickly becomes messy. Boundary between `Interval` and `OffsetDateTime` is hard to cross. `Instant.MIN.plusDays(1)` is relatively easy. 400 years forward, 1 day forward, 400 years backward, done. On the other hand, something like `OffsetLocalDate.MIN.plusDays(DAYS.between(LocalDate.MIN, LocalDate.MAX) + 1)` is _much_ harder to implement as _every_ shift will always end up outside `OffsetDateTime.MIN/MAX` bounds. 

I've also tried to experiment with `DateTimeFormatter.parseUnresolved`, but failed to find that helpful.

It seems that some `ExtendedOffsetDateTime` (using `ExtendedYear` internally (?), allowing values outside `Year.MIN/MAX`) + reimplementing all `TemporalAmount`-related calculations from scratch would be needed. That's too much for me.

To summarize: please consider this PR as ready.

**Comment by perceptron8** [comment]
@jodastephen Almost 4 months later, this PR is still ready to review... ;)
